### ISSUE TITLE
TestIsAddressCovered failing when localhost resolves to IPv6 address

### ISSUE THREAD

**Issue by piwaniuk** [lxc/lxd#8340]
# Required information

 * Distribution: Debian
 * Distribution version: bullseye
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.7.4
   * LXC version: 4.0.5
   * LXD version: 4.10
   * golang version: 1.15.6

# Issue description

When trying to build LXD from sources, the process failed at the unit test phase. After investigation it turns out the cause is that `net.LookupHost("localhost")` resolves to an IPv6 addres "::1" before "127.0.0.1".

# Steps to reproduce

 1. Download 4.10 source release
 2. Try to build it

# Information attached

 - build log snip [lxd-util-net_test.log](https://github.com/lxc/lxd/files/5825057/lxd-util-net_test.log)

**Comment by stgraber** [comment]
```
stgraber@castiana:~/data/code/lxc/lxd (lxc/master)$ go test -v -tags libsqlite3 ./lxd/util/
=== RUN   Test_CompareConfigsMismatch
=== RUN   Test_CompareConfigsMismatch/different_values_for_keys:_foo
=== RUN   Test_CompareConfigsMismatch/different_values_for_keys:_egg,_foo
--- PASS: Test_CompareConfigsMismatch (0.00s)
    --- PASS: Test_CompareConfigsMismatch/different_values_for_keys:_foo (0.00s)
    --- PASS: Test_CompareConfigsMismatch/different_values_for_keys:_egg,_foo (0.00s)
=== RUN   Test_CompareConfigs
--- PASS: Test_CompareConfigs (0.00s)
=== RUN   TestInMemoryNetwork
--- PASS: TestInMemoryNetwork (0.00s)
=== RUN   TestCanonicalNetworkAddress
=== RUN   TestCanonicalNetworkAddress/127.0.0.1
=== RUN   TestCanonicalNetworkAddress/foo.bar
=== RUN   TestCanonicalNetworkAddress/192.168.1.1:443
=== RUN   TestCanonicalNetworkAddress/f921:7358:4510:3fce:ac2e:844:2a35:54e
--- PASS: TestCanonicalNetworkAddress (0.00s)
    --- PASS: TestCanonicalNetworkAddress/127.0.0.1 (0.00s)
    --- PASS: TestCanonicalNetworkAddress/foo.bar (0.00s)
    --- PASS: TestCanonicalNetworkAddress/192.168.1.1:443 (0.00s)
    --- PASS: TestCanonicalNetworkAddress/f921:7358:4510:3fce:ac2e:844:2a35:54e (0.00s)
=== RUN   TestIsAddressCovered
=== RUN   TestIsAddressCovered/127.0.0.1:8443-127.0.0.1:8443
=== RUN   TestIsAddressCovered/garbage-127.0.0.1:8443
=== RUN   TestIsAddressCovered/127.0.0.1:8444-garbage
=== RUN   TestIsAddressCovered/127.0.0.1:8444-127.0.0.1:8443
=== RUN   TestIsAddressCovered/127.0.0.1:8443-0.0.0.0:8443
=== RUN   TestIsAddressCovered/[::1]:8443-0.0.0.0:8443
=== RUN   TestIsAddressCovered/:8443-0.0.0.0:8443
=== RUN   TestIsAddressCovered/127.0.0.1:8443-[::]:8443
=== RUN   TestIsAddressCovered/[::1]:8443-[::]:8443
=== RUN   TestIsAddressCovered/[::1]:8443-:8443
=== RUN   TestIsAddressCovered/:8443-[::]:8443
=== RUN   TestIsAddressCovered/0.0.0.0:8443-[::]:8443
=== RUN   TestIsAddressCovered/10.30.0.8:8443-[::]
=== RUN   TestIsAddressCovered/localhost:8443-127.0.0.1:8443
=== RUN   TestIsAddressCovered/127.0.0.1:8443-localhost:8443
=== RUN   TestIsAddressCovered/[::1]:8443-ip6-localhost:8443
--- PASS: TestIsAddressCovered (0.01s)
    --- PASS: TestIsAddressCovered/127.0.0.1:8443-127.0.0.1:8443 (0.00s)
    --- PASS: TestIsAddressCovered/garbage-127.0.0.1:8443 (0.01s)
    --- PASS: TestIsAddressCovered/127.0.0.1:8444-garbage (0.00s)
    --- PASS: TestIsAddressCovered/127.0.0.1:8444-127.0.0.1:8443 (0.00s)
    --- PASS: TestIsAddressCovered/127.0.0.1:8443-0.0.0.0:8443 (0.00s)
    --- PASS: TestIsAddressCovered/[::1]:8443-0.0.0.0:8443 (0.00s)
    --- PASS: TestIsAddressCovered/:8443-0.0.0.0:8443 (0.00s)
    --- PASS: TestIsAddressCovered/127.0.0.1:8443-[::]:8443 (0.00s)
    --- PASS: TestIsAddressCovered/[::1]:8443-[::]:8443 (0.00s)
    --- PASS: TestIsAddressCovered/[::1]:8443-:8443 (0.00s)
    --- PASS: TestIsAddressCovered/:8443-[::]:8443 (0.00s)
    --- PASS: TestIsAddressCovered/0.0.0.0:8443-[::]:8443 (0.00s)
    --- PASS: TestIsAddressCovered/10.30.0.8:8443-[::] (0.00s)
    --- PASS: TestIsAddressCovered/localhost:8443-127.0.0.1:8443 (0.00s)
    --- PASS: TestIsAddressCovered/127.0.0.1:8443-localhost:8443 (0.00s)
    --- PASS: TestIsAddressCovered/[::1]:8443-ip6-localhost:8443 (0.00s)
=== RUN   TestListenImplicitIPv6Wildcard
--- PASS: TestListenImplicitIPv6Wildcard (0.00s)
PASS
ok  	_/home/stgraber/data/code/lxc/lxd/lxd/util	0.015s

stgraber@castiana:~/data/code/lxc/lxd (lxc/master)$ getent hosts localhost
::1             localhost
```

So maybe more at play here?
What's in your `/etc/hosts` exactly?

**Comment by piwaniuk** [comment]
Here is my `/etc/hosts`:

```
127.0.0.1	localhost
::1		localhost ip6-localhost ip6-loopback
ff02::1		ip6-allnodes
ff02::2		ip6-allrouters
```

`net.LookupHost` returns a list of addresses and in the mentioned function only the first element is taken:

https://github.com/lxc/lxd/blob/lxd-4.10/lxd/util/net.go#L188

On the machine I run the build, the lookup apparently resolves to `{"::1", "127.0.0.1"}`. When I changed the function to compare all elements of the lookup result, the test passed.

**Comment by stgraber** [comment]
Looks like you prepared a fix for this, are you going to send it as a pull request?
